components/style/selector_parser.rs

```diff
@@ -173,3 +173,26 @@ impl<T> PerPseudoElementMap<T> {
         self.entries.iter()
     }
 }
+
+/// Values for the :dir() pseudo class
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Direction {
+    /// left-to-right semantic directionality
+    Ltr,
+    /// right-to-left semantic directionality
+    Rtl,
+    /// Some other provided directionality value
+    Other(Box<[u16]>),
+}
+
+impl<'a> From<&'a Direction> for String {
```

Comment:
Given we don't use the utf16 impl anyway, this could be `Box<str>`, and there'd be no need to allocate a `String` for serialization.